### ISSUE TITLE
No such device or address @ rb_sysopen - /dev/tty

### ISSUE THREAD

**Issue by bryanp** [ioquatix/process-group#2]
Hit this error running specs for one of my projects on Github Actions:

```
1) Pakyow::ProcessManager#add starts the expected number of processes
   Got 0 failures and 2 other errors:
   1.1) Failure/Error: Process::Group.new
        Errno::ENXIO:
          No such device or address @ rb_sysopen - /dev/tty
        # /usr/local/rvm/gems/ruby-2.5.7/gems/process-terminal-0.1.0/lib/process/terminal/device.rb:27:in `initialize'
        # /usr/local/rvm/gems/ruby-2.5.7/gems/process-terminal-0.1.0/lib/process/terminal/device.rb:27:in `open'
        # /usr/local/rvm/gems/ruby-2.5.7/gems/process-terminal-0.1.0/lib/process/terminal/device.rb:27:in `open'
        # /usr/local/rvm/gems/ruby-2.5.7/gems/process-group-1.2.0/lib/process/group.rb:102:in `initialize'
        # ./spec/unit/process_manager_spec.rb:16:in `new'
        # ./spec/unit/process_manager_spec.rb:16:in `block (2 levels) in <top (required)>'
        # ./spec/unit/process_manager_spec.rb:3:in `block (2 levels) in <top (required)>'
```

It seems that `process-terminal` assumes `/dev/tty` will be available. Opening the issue here even though the issue may be downstream because upgrading from `process-group` v1.1.0 to v1.2.0 breaks my use case. 

FWIW this change in my code seems to avoid the issue:

* https://github.com/pakyow/pakyow/pull/332

But I wouldn't expect to have to handle this on my end.

You can see the failure in context [here](https://github.com/pakyow/pakyow/runs/297971250).

**Comment by ioquatix** [comment]
Sorry about this, I will check now what is going on.

**Comment by ioquatix** [comment]
Yes, in a way this broke the expectations of the public interface.

I'm planning to in the first instance allow `Process::Terminal::Device.open` to return nil if the path doesn't exist.

**Comment by ioquatix** [comment]
<img width="1086" alt="image" src="https://user-images.githubusercontent.com/30030/68633092-f0399d00-0533-11ea-895b-1f1a9c463e6b.png">

I wonder if integrating these gems was wrong.

`process-group` can deal with groups of processes. It was nice to say, in this process group, make this one the foreground process. But maybe it was a mistake. But the process group can change the scheduling of processes if it knows one is in the foreground. I didn't realise `/dev/tty` would not be available for some reason. It means the CI is not connected to a pty?

**Comment by bryanp** [comment]
That's my understanding, that in this case CI is not connected to a pty. Our builds work by running a GitHub Actions that spins up a Docker container on a Google Compute instance, then runs commands via `net-ssh` (see https://github.com/pakyow/ci-helpers). Some of our builds are resource-intensive and don't run correctly on GitHub Actions, so this lets us run builds on predictable hardware. But I digress.

I'm not 100% sure what the right answer is here, but I do think it should work if `/dev/tty` is missing. I also don't think it was wrong to integrate the gems. Foregrounding from `process-group` makes sense to me. Something else that comes up as I'm thinking about this is support on non-*nix systems. I tested `process-group` out on Windows at some earlier point and it seemed to work fine--would relying on `/dev/tty` being available break this support in some way?

**Comment by ioquatix** [comment]
TTY/PTY gets really hairy on windows. But basically it should just gracefully degrade.

**Comment by ioquatix** [comment]
If the process is directly connected to the terminal, you can just use `STDIN`. I think this is a sane default, but I can't get tests to pass on github actions.

**Comment by bryanp** [comment]
You're referring to Windows, right? If so I'm not too worried about it and hope I didn't cause too much of a distraction with my comment. My only thought was that the ultimate solution here should gracefully degrade, including on non-*nix platforms. And I think your original plan does this:

> I'm planning to in the first instance allow Process::Terminal::Device.open to return nil if the path doesn't exist.

**Comment by ioquatix** [comment]
Yeah.

**Comment by ioquatix** [comment]
Well, in order for a process to be promoted to foreground the stdin probably needs to be the input of the terminal.... so opening `/dev/tty` is kind of unnecessary because stdin should already be connected to the terminal... if not, there might be other problems. So, I think it's an okay default.

**Comment by ioquatix** [comment]
Okay, so there are two changes:

- https://github.com/socketry/process-terminal/blob/ff89119b02ea89e779425533420e374c420c0a76/lib/process/terminal/device.rb#L26-L32 is best effort, if the path isn't readable it fails and returns nil which means process-group skips trying to move processes to the foreground (affects/fixes process-group v1.2.0).

- https://github.com/socketry/process-terminal/blob/ff89119b02ea89e779425533420e374c420c0a76/lib/process/terminal/device.rb#L34-L39 is an optional constructor that can return nil if STDIN is not connected to a terminal. It's logical to me that this should be the default going forward (little/no overhead, affects/fixes process-group v1.2.1).

Sorry about this mess, I'm still probably on the fence about how this should work - pulling in FFI is a big dependency. I might try to get some of these C APIs exposed by Ruby itself.

**Comment by bryanp** [comment]
Thanks for your work on this!
